File: backend/app/schemas/tasks.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

TaskStatus = str | None
MarkStatus = Literal["reward", "punishment"] | None


class TaskCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Task name is required")
        return normalized


class TaskUpdateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Task name is required")
        return normalized


class TaskHistoryItem(BaseModel):
    date: date
    status: TaskStatus


class TaskListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    created_at: datetime
    archived_at: datetime | None
    is_pinned: bool
    pinned_at: datetime | None
    selected_date_status: TaskStatus
    history: list[TaskHistoryItem]


class TaskListResponse(BaseModel):
    items: list[TaskListItem]


class TaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    created_at: datetime
    archived_at: datetime | None
    is_pinned: bool
    pinned_at: datetime | None


class TaskDetailResponse(TaskRead):
    total_reward: int
    total_punishment: int


class TaskMarkUpsertRequest(BaseModel):
    status: MarkStatus


class TaskMarkResponse(BaseModel):
    task_id: UUID
    date: date
    status: MarkStatus

File: frontend/src/api/tasks.ts
import { apiClient } from "./client";

export type TaskStatus = "reward" | "punishment" | null;

export type TaskHistoryItem = {
  date: string;
  status: TaskStatus;
};

export type TaskListItem = {
  id: string;
  name: string;
  created_at: string;
  archived_at: string | null;
  is_pinned: boolean;
  pinned_at: string | null;
  selected_date_status: TaskStatus;
  history: TaskHistoryItem[];
};

export type TaskListResponse = {
  items: TaskListItem[];
};

export type TaskRead = {
  id: string;
  name: string;
  created_at: string;
  archived_at: string | null;
  is_pinned: boolean;
  pinned_at: string | null;
};

export type TaskDetail = TaskRead & {
  total_reward: number;
  total_punishment: number;
};

export type TaskMarkResponse = {
  task_id: string;
  date: string;
  status: TaskStatus;
};

export const tasksApi = {
  list: (selectedDate: string) => apiClient.get<TaskListResponse>(`/api/tasks?date=${selectedDate}`),
  create: (name: string) => apiClient.post<TaskRead>("/api/tasks", { name }),
  update: (taskId: string, name: string) => apiClient.patch<TaskRead>(`/api/tasks/${taskId}`, { name }),
  details: (taskId: string) => apiClient.get<TaskDetail>(`/api/tasks/${taskId}`),
  archive: (taskId: string) => apiClient.post<TaskRead>(`/api/tasks/${taskId}/archive`),
  restore: (taskId: string) => apiClient.post<TaskRead>(`/api/tasks/${taskId}/restore`),
  archived: () => apiClient.get<TaskRead[]>("/api/tasks/archived"),
  pin: (taskId: string) => apiClient.post<TaskRead>(`/api/tasks/${taskId}/pin`),
  unpin: (taskId: string) => apiClient.post<TaskRead>(`/api/tasks/${taskId}/unpin`),
  setMark: (taskId: string, markDate: string, status: TaskStatus) =>
    apiClient.put<TaskMarkResponse>(`/api/tasks/${taskId}/marks/${markDate}`, { status }),
};
